src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3164,29 +3161,43 @@ void checkLoopCondition()
                 }
                 else if (userDefinedInfo != null)
                 {
-                    // PROTOTYPE(dataflow): Handle "user-defined" loops
-                    throw new NotImplementedException();
+                    Debug.Assert(_forToLoopBinaryOperatorLeftOperand == null);
+                    Debug.Assert(_forToLoopBinaryOperatorRightOperand == null);
 
-                    //    ' Generate If(positiveFlag, controlVariable <= limit, controlVariable >= limit)
-                    //AddPlaceholderReplacement(operatorsOpt.LeftOperandPlaceholder, controlVariable.MakeRValue())
-                    //AddPlaceholderReplacement(operatorsOpt.RightOperandPlaceholder, limit)
+                    // Generate If(positiveFlag, controlVariable <= limit, controlVariable >= limit)
 
-                    //Dim result = MakeTernaryConditionalExpression(operatorsOpt.Syntax,
-                    //                                                   New BoundLocal(operatorsOpt.Syntax,
-                    //                                                                  positiveFlag,
-                    //                                                                  isLValue:= False,
-                    //                                                                  type:= positiveFlag.Type),
-                    //                                                   VisitExpressionNode(operatorsOpt.LessThanOrEqual),
-                    //                                                   VisitExpressionNode(operatorsOpt.GreaterThanOrEqual))
+                    // Spill control variable reference, we are going to have branches here.
+                    int captureId = _availableCaptureId++;
+                    IOperation controlVariableReferenceforCondition = getLoopControlVariableReference(forceImplicit: true, captureId); // Yes we are going to reevaluate it again
```

Comment:
>Referencefor [](start = 46, length = 12)

`ReferenceFor`

Reply:
> ReferenceFor 

Will adjust the name in the next PR

---
In reply to: [187707603](https://github.com/dotnet/roslyn/pull/26771#discussion_r187707603) [](ancestors = 187707603)